### DISCUSSION TITLE
Vehicle repair and replacement - future directions

### DISCUSSION THREAD

**Discussion by I-am-Erk** [CleverRaven/Cataclysm-DDA#59046] [General]
#59001 has opened up some discussion on vehicle repair that has been brewing for a while on discord but should be formalized somewhere. This PR aims to make factory frames distinguishable from homemade ones, which has two major gameplay roles.

1. It becomes less easy to just smash everything with a car, because your car frame becomes increasingly damaged and cannot be easily repaired
2. It becomes much more complex to turn a small vehicle into a large one by welding new frames on, because the new frames are heavier and more brittle than the others.

There are some future directions here getting brought up in the PR, and I think they should be discussed elsewhere.

## Replacing commercial parts with other commercial parts

This is a thing that should be possible, but without supply lines it is not easy. You can't replace the front end of one car with the rear end of a totally different car. I suggest the following stages of code modification:
1. Add model, and colour as largely cosmetic features of vehicles in game, allowing us to distinguish two otherwise identical sedans
2. Track what model (and colour if appropriate) a piece comes from, and what part, so 'front left corner frame' and 'Coyote Windwaker' are bits of metadata we attach to an otherwise normal 'frame' item
3. to replace a commercial part you need a piece with the matching metadata.
4. Later, skills and tools can let parts 'wobble', so you can use a front left corner frame from any car in a 'coyote windwaker' with a small penalty

Note that in this case, replacing a part should still not bring it quite up to perfect. This shouldn't decay the durability as much as repair would, but it should still decay slightly. Eventually, your car will just have had too many accidents and you can't ship-of-theseus it back to commercial spec. you'll have to go to a new vehicle or use jury rigged parts.

## Faults instead of durability
Per Kevin: something that has come up in thinking about this is that instead of unit less damage accumulating on the parts, maybe we do want to induce faults instead which might represent things a little bit better. So as as you encounter damage (mostly crash events) your individual frames develop "cracks" and "split welds" and and accumulate metal fatigue and it's hitting the whole car at once to some extent. This has the outcome we're looking for where instead of just having one frame that repeatedly breaks and gets replaced you have an accumulation of damage across the whole frame.  This leads to a natural trade off where at some point it becomes obvious that the frame is not worth salvaging and you just want to move all your parts to a new frame

- I will add that this would be interesting if some faults were tracked on a vehicle level, like "out of alignment", and replacing parts has no impact on them. In that example you'd need the skills and equipment to fix the alignment. Some such faults may be prohibitively difficult to repair in an apocalypse.

## Add weight allowance to frame parts
Currently jury-rigged frames shouldn't be too much different from commercial, making this less of a nerf than it appears. They should be heavier, but similarly durable. However, an important next step is to ensure frames can bend and be damaged by carry weight, and jury rigged frames should wear our easier this way.

## Sanity check 'transferred' damage
If I am not mistaken, we already track shock damage to other vehicle parts and characters in the vehicle during an impact. However I don't think we're very granular about this. Commercial frames should absorb much more damage, keeping other parts of the vehicle and its content from getting harmed. Jury rigged frames should transfer a huge amount of damage, meaning an impact shares damage across all other vehicle parts and, significantly, *to the passengers*. This is a major change that would allow us to make more jury rigged frames that aren't necessarily even worse than commercial frames, but have a big tradeoff. Your car will be less damaged by ramming a hulk... *you* will not be. This is a great change to vehicles that can make them much more combat-balanced. Commercial vehicles are single-use crumple pots, then, and if you build your own 50's style body frame, it is more durable in crashes but you'll get quite bashed up yourself.

**Comment by esotericist** [comment]
once we've got the proper ability to distinguish between factory frames and crafted frames, being able to explore better weight ratios (esp with complex materials) for factory vehicles would allow us to have variation in fuel economy aside from just which engine is involved and what size of vehicle.

we still need to fix in-game distances before that matters, of course.

**Comment by I-am-Erk** [comment]
Quoting @PatrikLundell from the associated PR, to redirect conversation here:
> Well, "too many" is approximately one gotcha graboid. Any collision with those buggers at marching speed will require repair and probably permanently damaged parts, with shock absorbers being fairly good at protecting the interior.
> 
> I think it's possible to do limited zombie plowing at lower speeds with limited damage, but haven't dared to try it after the permanent damage introduction. I wouldn't use regular frames for anything that's supposed to survive a collision with anything (they can sometimes hit shrubs without taking damage).
> 
> I disagree with @kevingranade on repairing a custom vehicle. The whole reason I want a custom vehicle is because pre cataclysm vehicles aren't made for the post cataclysm world. Compression zones etc. are great for protecting lives in a collision, but when the creatures you collide with are undead you don't want to protect them from harm, and you definitely don't want to damage your vehicle doing it. If repairing the custom vehicle becomes so expensive that you'd have to make a new one, then I'd do that rather than use a damaged standard vehicle that falls apart at the next collision with something hostile, leaving you stranded and surrounded. If a security van was up to the task I wouldn't create a custom vehicle (although I'd still lament that it's too bulky and unwieldy), and even if it was, it will fall apart due to damage since you can't repair it any longer. Good luck finding two suitable vehicles in pristine, or even good condition.

**Reply by I-am-Erk** [reply]
There are two separate things to discuss here that I can see.

1. Graboids shouldn't pop up as a gotcha for your car so often. That is fine, we can adjust that.
2. You want a custom vehicle that is appropriate to an apocalypse. Great! So do I. All our goals here are in line, and adding more ways to distinguish a consumer passenger car from an APC or a custom-built deathmobile is on point. The changes described don't prevent or limit that, they make it more desirable by distinguishing your zombie murder machine from a basic sedan you grabbed from the roadside.

**Reply by esotericist** [reply]
yeah, ground ambush enemies are an excessively punitive threat to vehicles in a weird way atm, fixing that would be good. (at least we don't have to worry about burrowing spiders teleporting into your vehicle anymore)

**Comment by keampe** [comment]
Writing this here but responding to the PR (don't want to clutter up the PR).

Not sure how to quote from the PR. @kevingranade : The reason I linked the desire to fix a car from basic materials between a hobbyist car restorer and a survivor is that the survivor may not have a choice. As you said - cars are precious. If your survivor has access to many cars then ya, switch vehicles when one becomes damaged. I've been known to dump cars (in game) that run out of gas. But sooner or later that stops working. When you have access to only one car - you fix it however you can. If you need to fabricate a frame, that's what you do. But you are quite correct - if you have other options then you have other priorities as a survivor.

   - Shane

**Reply by I-am-Erk** [reply]
cars aren't really that precious in the game. even as it stands, they're pretty commonplace, and the desired direction is to have a lot more of them with the problem being more a matter of how to get them out of traffic jams than anything else, and then later, how to fuel them.

**Reply by esotericist** [reply]
*cars in general* are commonplace. a *functioning* vehicle that is in an accessible place, that you can make use of, that's the part that makes them precious.

**Reply by keampe** [reply]
I was replying directly to @kevingranade when he made the statement of cars being precious. As to the direction of the game - is that anywhere I can peruse? I was totally unaware of anything that states that direction is more vehicles in working order.

   - Shane

**Reply by esotericist** [reply]
a lot of this happens in discord discussions.

**Reply by I-am-Erk** [reply]
Should be noted that a car in working order is less likely, but the models we are describing are models where, if you clear the zombies out, there are a lot of cars you can harvest to get the parts that you've ruined on your other one. Even if that means stripping it to the chassis and using that.

**Reply by keampe** [reply]
OK, that actually makes a lot of sense. I can easily imagine a panic causing a traffic jam which causes a demolition derby as people use their cars to GTFO. Scary as hell, and very plausible.

    - Shane

**Comment by keampe** [comment]
Factory parts.... Don't cars have an ID? If you take the ID of the model of vehicle and the xy position of the frame and attach that to the frame as it's being removed then you'd have the factory part, wouldn't you? You could use this for other parts, too, like quarterpanels.

Frames found outside of a car could be renamed to "Frame Parts" that get turned into a frame when installed and if subsequently removed they'd get the same metadata as any other frame. Frame Parts could be fabricated by the player at very high skill, likely locked behind proficiencies.

Frames could also have an "Adapt frame and install" option which would allow you to take one frame and adapt it to a different vehicle/position in vehicle with a hit to quality. It becomes a "best you could do" scenario.

   - Shane

**Comment by natsirt721** [comment]
What about using this as a driver towards multi-tile parts? You could have a 4x5 'sedan frame' that the pre-apoc. sedan is built on top of, and the player can repair and mend this to a degree, but eventually its going to be too fragile to be safe and/or too costly to repair. With the right balance of durability/accumulated wear/repair cost, this can drive the desired behavior. This way, you can give vehicles identity without having to have unique parts for the front/center/rear/etc. for each model of vehicle.

While not really any different conceptually from the related PR, it would help to separate the concept of 'base vehicle frame' vs 'applique/custom parts' from the player perspective. This also encourages the player to select an appropriately sized/shape vehicle to begin customization on, rather than starting with anything and adding tons of custom frames to achieve the right geometry. If desired, you could limit the type/number of custom frame parts to prevent arbitrarily large vehicles being created from small frames.

With the right skill, material, and tooling, players would eventually be able to create their own multitile frames, but these would be smaller and definitely require significant investments in tooling to achieve and high skill and unique proficiencies. Material would probably not be a limiting factor at this point.

Makes handling distributed damage easier as well.  Damage passes from the environment to the external part, then into the frame, then distributed damage/shock damage into attached parts. Or something like that.

This applies to car-sized vehicles and up - the tile-by-tile approach should be appropriate for smaller vehicles and vehicles made with wooden parts or lighter frames.

**Comment by Tharn** [comment]
The welder you have access to is not any less capable than a welder in a garage or machine shop. We get metal pipes, various steel sheets, various steels and aluminium. If you're aiming to distinguish "real" car frames from homebuilt ones, then give the survivor the ability to build and repair actual "real" frames with their tools. Make it take longer like it works for higher-grade melee weapons currently. But the notion that anything the survivor can do should necessarily be of lesser quality than a "full" commercial repair is flawed.

Bent or cracked pieces of factory frames are regularly cut out and replaced and welded into place by car technicians in the modern day and the result _does not weaken the car_. It just has to be properly executed and be the right material. I am firmly against the idea that cars should degrade irreparably. That's just more tedium for the player and it will make death mobiles a bad idea. (You could argue here that they are silly and they _should_ be a bad idea - but they are not. They are massively cool and a defining feature of what makes CDDA great. Leeway needs to be given to allow this feature to continue living.) As soon as the frame is degraded far enough, we're supposed to either abandon it or transfer the car parts and items worth of a _large military APC_ to another frame. Who's going to do that? Ever?

As you can see by the issue linked by the OP, players _love_ (sarcasm) the idea of not being able to repair their cars properly.

**Comment by JReese3** [comment]
Just want to point out that some crazy people do header conversions to allow non stock engines in other models.  I personally hate that this is even an idea.
